File: revision/singleNumber.py
from operator import xor
from re import X
from typing import List
def singleNumber(nums:List[int])->int:
    '''
    Approach 1: TC: O(N), SC: O(N)
    Input: nums = [4,1,2,1,2]
    Output: 4
    return 2*sum(set(nums))- sum(nums) 
    2 *(7) -   (10)
    14 - 10 = 4
    '''
    # return 2 * sum(set(nums)) - sum(nums)
    '''
    Approach 2 : Using XOR 
    Input: nums = [4,1,2,1,2]
    Output: 4
    4 ^ 1 ^ 2 ^ 1 ^ 2
    4 ^ 1 ^ 1 ^ 2 ^ 2
    4 ^ 0 ^ 0 
    4 ^ 0 = 4
    '''
    res = 0
    for x in nums:
        res ^= x
    return res
    
nums = list(map(int, input().split()))
print(singleNumber(nums))

File: revision/singleNumber.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
using namespace std;
void input(vector<int>&nums){
   for(auto &i : nums){
       cin >> i;
   }
}
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
int singleNumber(vector<int>&nums){
    //Using hash set
    // int sum_hashset = 0, sum= 0;
    // unordered_set<int>myhashset;
    // for(auto i: nums){
    //     myhashset.insert(i);
    // }
    // for(auto i: myhashset){
    //     sum_hashset+=i;        
    // }
    // for(auto i: nums){
    //     sum+=i;
    // }
    // return 2 * sum_hashset - sum;

    //Using XOR
    int res = 0;
    for(auto i : nums){
        res ^= i;
    }
    return res;
}
int main()
{
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    cout << singleNumber(nums)<<endl;
    return 0;
}